### DISCUSSION TITLE
Support for "new ()" and "default" as a reference to a default static instance

### DISCUSSION THREAD

**Discussion by Xyncgas** [dotnet/csharplang#5666] [Language Ideas] [answered]
I assume that new() creates a new instance for the object, where as default for reference type is null and for value type is its default.
say I initialize an object using new() or default when I declare them, I might leave it as is for a while until I do something with it, therefore I propose that until I do something with it the object should have a reference to a readonly static instance of that type, making the memory footprint smaller, and when I actually do something to the variable it then initialize the instance lazily.

Therefore having a 'default' state where objects is going to be, and also giving people an alternative to declare their variables to reference to the default state, instead of null.

It's kinda useful for (saving space) when wondering how much customers are there at a certain floor, instead of leaving the List of Customers as null I would normally initialize it with new() so it can tell me there is 0 customer if I hadn't added any, but if 7 out of 10 floors are empty and each ```floor``` has an instance of ```List<Customer>``` that was initialized with new() when declaring that's 7 different instance that's the same thing

Today, I can do it myself, as I've considered several options :
```Lazy<T>``` : However, it still generates a new instance upon request (Memory +)
```Custom getter & setter, if setter never called getter returns a reference to static instance``` : It works I've got to use it on every variable I declared or I can write a custom type to support this behavior (Code +)

Therefore I think it's fitting to use it in the language, just let new() point to a default static instance of that type that has the get set behavior as desired generated by the compiler. We can go one step further if we want to do this with the ```default``` keyword too

**Comment by HaloFour** [comment]
`new()` with a reference type means to allocate a new instance of that type.  If you didn't intend to create an instance of that type and you want to defer the allocation of an instance, don't use `new()`.

**Comment by Richiban** [comment, marked as answer]
Now that we have static abstract interface members, and we're hopefully going to get this-types at some point soon, an interface such as this should be possible:

```cs
public interface IDefaultable
{
    static abstract This Default { get; }
}

public class C
{
    public void M<T>() where T : IDefaultable<T>
    {
        var x = T.Default;
    }
}
```

Hopefully that interface (or something like it) will make it into the BCL someday, because I also frequently wish for a way to ask for a default value for a type in C# that is not necessarily null or zero.

**Reply by Xyncgas** [reply]
Well anything that modifies the static reference should create an instance of it, therefore I recommand changing it to readonly and implement this interface with record types

**Reply by DavidArno** [reply]
@Richiban,

I'm unsure of the benefit of the this type here. What advantage does that offer over doing this way?
```c#
public interface IDefaultable<T>
{
    public static abstract T Default { get; }
}
```

**Reply by Richiban** [reply]
Well, the this-type isn't _necessary_, but it definitely has a benefit. I'm not sure whether the language team has settled on the usage of CRTPs for solving this problem but it's always struck me as something of a hack in the interfaces of `IEquatable<T>` etc.

Without the this-type there's nothing to prevent the user from writing:

```cs
public class Point : IDefaultable<string>// <--- This is nonsense but perfectly legal
{
    public int X { get; init; }
    public int Y { get; init; }

   public static string Default => "";
}
```

**Comment by huoyaoyuan** [comment]
`new` and `default` have their dedicated instruction in IL, special meaning and CLR. `new` means a GC allocation. `default` means all bits zero.  Every C# users are already aware of this.

**Reply by Xyncgas** [reply]
Can't we change default to mean something else than all bits 0
like a default state for object

**Reply by HaloFour** [reply]
Not easily.  It would require quite a bit of runtime support.

**Comment by TahirAhmadov** [comment]
The solution you're looking for, I think, is `Lazy<T>`.
